pyro/contrib/forecast/forecaster.py

```diff
@@ -283,3 +284,80 @@ def forward(self, data, covariates, num_samples):
                     return self.model(data, covariates)
+
+
+class HMCForecaster(nn.Module):
+    """
+    Forecaster for a :class:`ForecastingModel` using Hamiltonian Monte Carlo.
+
+    On initialization, this will run :class:`~pyro.infer.mcmc.nuts.NUTS` sampler
+    to get posterior samples of the model.
+
+    After construction, this can be called to generate sample forecasts.
+
+    :param ForecastingModel model: A forecasting model subclass instance.
+    :param data: A tensor dataset with time dimension -2.
+    :type data: ~torch.Tensor
+    :param covariates: A tensor of covariates with time dimension -2.
+        For models not using covariates, pass a shaped empty tensor
+        ``torch.empty(duration, 0)``.
+    :type covariates: ~torch.Tensor
+
+    :param int num_warmup: number of MCMC warmup steps.
+    :param int num_samples: number of MCMC samples.
+    :param int num_chains: number of parallel MCMC chains.
+    :param bool dense_mass: a flag to decide if mass matrix is dense or diagonal.
+        Default to False.
+    :param bool jit_compile: whether to use the PyTorch JIT to trace the log
+        density computation, and use this optimized executable trace in the
+        integrator. Default to False.
+    :param int max_tree_depth: Max depth of the binary tree created during the doubling
+        scheme of :class:`~pyro.infer.mcmc.nuts.NUTS` sampler. Default to 10.
+    """
+    def __init__(self, model, data, covariates=None, *,
+                 num_warmup=1000, num_samples=1000, num_chains=1,
+                 dense_mass=False, jit_compile=False, max_tree_depth=10):
+        assert data.size(-2) == covariates.size(-2)
+        super().__init__()
+        self.model = model
+        max_plate_nesting = _guess_max_plate_nesting(model, (data, covariates), {})
+        self.max_plate_nesting = max(max_plate_nesting, 1)  # force a time plate
+
+        kernel = NUTS(model, full_mass=dense_mass, jit_compile=jit_compile, ignore_jit_warnings=True,
+                      max_tree_depth=max_tree_depth, max_plate_nesting=max_plate_nesting)
+        mcmc = MCMC(kernel, warmup_steps=num_warmup, num_samples=num_samples, num_chains=num_chains)
+        mcmc.run(data, covariates)
+        # conditions to compute rhat
+        if (num_chains == 1 and num_samples >= 4) or (num_chains > 1 and num_samples >= 2):
+            mcmc.summary()
+
+        # inspect the model with particles plate = 1, so that we can reshape samples to
+        # add any missing plate dim in front.
+        with poutine.trace() as tr:
+            with pyro.plate("particles", 1, dim=-self.max_plate_nesting - 1):
+                model(data, covariates)
+
+        self._trace = tr.trace
+        self._samples = mcmc.get_samples()
+        self._num_samples = num_samples * num_chains
+        for name, node in list(self._trace.nodes.items()):
+            if name not in self._samples:
+                del self._trace.nodes[name]
```

Comment:
what is this doing?

Reply:
as I understand this is garbage collecting param statements to save memory

Reply:
In addition, this is to remove sites such as residuals, which should not be replayed in forward method.